hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +48,130 @@
+         * <p>
+         * No-op if less than {@link #MC_EVENTS_WINDOW_MILLIS} millis passed since the last cleanup. This avoids too frequent
+         * unnecessary iterations on the {@link #lastPollRecordPerMCAddress} entries.
+         */
+        private void cleanUpLastAccessRecords() {
+            long now = clock.getAsLong();
+            if (!isOutOfTimeWindow(now, lastCleanupTimestamp)) {
+                return;
+            }
+            lastPollRecordPerMCAddress.entrySet().removeIf(entry -> isOutOfTimeWindow(now, entry.getValue().lastAccessTime));
+            lastCleanupTimestamp = now;
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
+            lastPollRecordPerMCAddress.clear();
+        }
+
+        /**
+         * @param mcRemoteAddr the address of the calling MC instance.
+         * @return the events which were added to the queue since this MC last polled, or all known events if the MC polls for the
+         * first time.
+         */
+        public List<Event> pollMCEvents(Address mcRemoteAddr) {
+            lastMCEventsPollMillis = clock.getAsLong();
+            LastPollRecord lastPollRecord = lastPollRecordPerMCAddress.get(mcRemoteAddr);
+            long sequence;
+            if (lastPollRecord == null) {
+                sequence = 0;
+            } else {
+                sequence = lastPollRecord.nextSequence;
+            }
+            ConcurrentArrayRingbuffer.RingbufferSlice<Event> slice = mcEvents.copyFrom(sequence);
+            lastPollRecordPerMCAddress.put(mcRemoteAddr, new LastPollRecord(lastMCEventsPollMillis, slice.nextSequence()));
+            return slice.elements();
+        }
+    }
+
     public static final String SERVICE_NAME = "hz:core:managementCenterService";
 
     private static final int MIN_EVENT_QUEUE_CAPACITY = 1000;
     private static final int EXECUTOR_QUEUE_CAPACITY_PER_THREAD = 1000;
     private static final long TMS_CACHE_TIMEOUT_NANOS = TimeUnit.SECONDS.toNanos(1);
-    private static final long MC_EVENTS_WINDOW_NANOS = TimeUnit.SECONDS.toNanos(30);
 
     private final HazelcastInstanceImpl instance;
     private final ILogger logger;
 
     private final AtomicReference<String> tmsJson = new AtomicReference<>();
     private final TimedMemberStateFactory tmsFactory;
     private final AtomicBoolean tmsFactoryInitialized = new AtomicBoolean(false);
-    private final BlockingQueue<Event> mcEvents;
     private final ConsoleCommandHandler commandHandler;
     private final ClientBwListConfigHandler bwListConfigHandler;
+    private final MCEventStore eventStore;
 
     private volatile ManagementCenterEventListener eventListener;
     private volatile String lastMCConfigETag;
     private volatile long lastTMSUpdateNanos;
-    private volatile long lastMCEventsPollNanos = System.nanoTime();
 
     public ManagementCenterService(HazelcastInstanceImpl instance) {
+        this(instance, Clock::currentTimeMillis);
```

Comment:
Monotonic clock (`System#nanoTime()`) should be used now in `MCEventStore` since it doesn't deal with event time anymore.

Reply:
Done.